In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1.20.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

In [2]:
CLIENT_ID = 'WVVDZRIEPUF43TCVXTB1T3JEHNC41AAS3YJAO4SRIQCE14U2' # your Foursquare ID
CLIENT_SECRET = 'CMB21XECJFAFCAZ5CDMP53AWTZBNUACTEBLGHSTBLPUQ2JFX' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: WVVDZRIEPUF43TCVXTB1T3JEHNC41AAS3YJAO4SRIQCE14U2
CLIENT_SECRET:CMB21XECJFAFCAZ5CDMP53AWTZBNUACTEBLGHSTBLPUQ2JFX


### Lets Search for Cafes near Eiffel Tower

In [3]:
address = 'Eiffel Tower, Paris'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

48.858260200000004 2.2944990543196795


In [4]:
search_query = 'Coffee'
radius = 1000
print(search_query + ' .... OK!')

Coffee .... OK!


In [5]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=WVVDZRIEPUF43TCVXTB1T3JEHNC41AAS3YJAO4SRIQCE14U2&client_secret=CMB21XECJFAFCAZ5CDMP53AWTZBNUACTEBLGHSTBLPUQ2JFX&ll=48.858260200000004,2.2944990543196795&v=20180604&query=Coffee&radius=1000&limit=30'

In [6]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e223fa902a172002870b0b9'},
 'response': {'venues': [{'id': '4f5f73bfe5e8576b5047e1aa',
    'name': 'Coffee & Cigarettes',
    'location': {'address': '3 avenue Octave Greard',
     'lat': 48.8564422792504,
     'lng': 2.2933316230773926,
     'labeledLatLngs': [{'label': 'display',
       'lat': 48.8564422792504,
       'lng': 2.2933316230773926}],
     'distance': 219,
     'postalCode': '75007',
     'cc': 'FR',
     'city': 'Paris',
     'state': 'Île-de-France',
     'country': 'France',
     'formattedAddress': ['3 avenue Octave Greard', '75007 Paris', 'France']},
    'categories': [{'id': '4bf58dd8d48988d123951735',
      'name': 'Smoke Shop',
      'pluralName': 'Smoke Shops',
      'shortName': 'Smoke Shop',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/tobacco_',
       'suffix': '.png'},
      'primary': True}],
    'referralId': 'v-1579303133',
    'hasPerk': False},
   {'id': '5220825e11d2a2b972c0b5ac',
    'nam

In [7]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories  hasPerk  \
0  [{'id': '4bf58dd8d48988d123951735', 'name': 'S...    False   
1  [{'id': '4bf58dd8d48988d1c5941735', 'name': 'S...    False   
2  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
3  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
4  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   

                         id             location.address location.cc  \
0  4f5f73bfe5e8576b5047e1aa       3 avenue Octave Greard          FR   
1  5220825e11d2a2b972c0b5ac         41 avenue de Suffren          FR   
2  51d592e8498e77528472756f  67 avenue de la Bourdonnais          FR   
3  5a609ae84940bc265a72169b       90 rue Saint Dominique          FR   
4  4df1f8e13151247c51af4f21       90 rue Saint-Dominique          FR   

  location.city location.country  location.distance  \
0         Paris           France                219   
1         Paris           France                453   
2         Paris           France                542   
3         Paris           France                827   
4         Paris           France                851   

                           location.formattedAddress  \
0      [3 avenue Octave Greard, 75007 Paris, France]   
1        [41 avenue de Suffren, 75007 Paris, France]   
2  [67 avenue de la Bourdonnais, 75007 Paris, Fra...   
3      [90 rue Saint Dominique, 75007 Paris, France]   
4      [90 rue Saint-Dominique, 75015 Paris, France]   

                             location.labeledLatLngs  location.lat  \
0  [{'label': 'display', 'lat': 48.8564422792504,...     48.856442   
1  [{'label': 'display', 'lat': 48.8543382227668,...     48.854338   
2  [{'label': 'display', 'lat': 48.85673894236335...     48.856739   
3  [{'label': 'display', 'lat': 48.85938, 'lng': ...     48.859380   
4  [{'label': 'display', 'lat': 48.85943355462305...     48.859434   

   location.lng location.postalCode location.state                 name  \
0      2.293332               75007  Île-de-France  Coffee & Cigarettes   
1      2.296151               75007  Île-de-France          Coffee & Co   
2      2.301542               75007  Île-de-France          Coffee Shop   
3      2.305670               75007  Île-de-France            Starbucks   
4      2.305984               75015  Île-de-France            Starbucks   

     referralId  
0  v-1579303133  
1  v-1579303133  
2  v-1579303133  
3  v-1579303133  
4  v-1579303133

In [8]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name      categories                        address  cc  \
0  Coffee & Cigarettes      Smoke Shop         3 avenue Octave Greard  FR   
1          Coffee & Co  Sandwich Place           41 avenue de Suffren  FR   
2          Coffee Shop            Café    67 avenue de la Bourdonnais  FR   
3            Starbucks     Coffee Shop         90 rue Saint Dominique  FR   
4            Starbucks     Coffee Shop         90 rue Saint-Dominique  FR   
5            Starbucks     Coffee Shop  66 avenue de la Motte-Picquet  FR   
6         O Coffeeshop     Coffee Shop              23 rue de Lourmel  FR   

    city country  distance                                   formattedAddress  \
0  Paris  France       219      [3 avenue Octave Greard, 75007 Paris, France]   
1  Paris  France       453        [41 avenue de Suffren, 75007 Paris, France]   
2  Paris  France       542  [67 avenue de la Bourdonnais, 75007 Paris, Fra...   
3  Paris  France       827      [90 rue Saint Dominique, 75007 Paris, France]   
4  Paris  France       851      [90 rue Saint-Dominique, 75015 Paris, France]   
5  Paris  France      1008  [66 avenue de la Motte-Picquet, 75015 Paris, F...   
6  Paris  France      1049           [23 rue de Lourmel, 75015 Paris, France]   

                                      labeledLatLngs        lat       lng  \
0  [{'label': 'display', 'lat': 48.8564422792504,...  48.856442  2.293332   
1  [{'label': 'display', 'lat': 48.8543382227668,...  48.854338  2.296151   
2  [{'label': 'display', 'lat': 48.85673894236335...  48.856739  2.301542   
3  [{'label': 'display', 'lat': 48.85938, 'lng': ...  48.859380  2.305670   
4  [{'label': 'display', 'lat': 48.85943355462305...  48.859434  2.305984   
5  [{'label': 'display', 'lat': 48.84952043604051...  48.849520  2.298146   
6  [{'label': 'display', 'lat': 48.84908977470434...  48.849090  2.291148   

  postalCode          state                        id  
0      75007  Île-de-France  4f5f73bfe5e8576b5047e1aa  
1      75007  Île-de-France  5220825e11d2a2b972c0b5ac  
2      75007  Île-de-France  51d592e8498e77528472756f  
3      75007  Île-de-France  5a609ae84940bc265a72169b  
4      75015  Île-de-France  4df1f8e13151247c51af4f21  
5      75015  Île-de-France  4d4e7252a0b0a093c4ea4d83  
6      75015  Île-de-France  585cdf1cea1c0d03328885cc

In [9]:
dataframe_filtered.name

0    Coffee & Cigarettes
1            Coffee & Co
2            Coffee Shop
3              Starbucks
4              Starbucks
5              Starbucks
6           O Coffeeshop
Name: name, dtype: object

In [ ]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Eiffel Tower

# add a red circle marker to represent the Eiffel Tower
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Eiffel Tower',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Coffee places as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

## Explore a particular venue

### Now we will explore Starbucks which is at a distance of 851 metres from us and check its rating.

In [11]:
venue_id = '4df1f8e13151247c51af4f21' # ID of Starbucks
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/4df1f8e13151247c51af4f21?client_id=WVVDZRIEPUF43TCVXTB1T3JEHNC41AAS3YJAO4SRIQCE14U2&client_secret=CMB21XECJFAFCAZ5CDMP53AWTZBNUACTEBLGHSTBLPUQ2JFX&v=20180604'

In [12]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'price', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'page', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'hours', 'popular', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto', 'colors'])


{'id': '4df1f8e13151247c51af4f21',
 'name': 'Starbucks',
 'contact': {'phone': '+33145511257',
  'formattedPhone': '+33 1 45 51 12 57',
  'twitter': 'starbucksfrance',
  'instagram': 'starbucksfrance',
  'facebook': '188336880869',
  'facebookUsername': 'StarbucksFrance',
  'facebookName': 'Starbucks France'},
 'location': {'address': '90 rue Saint-Dominique',
  'lat': 48.85943355462305,
  'lng': 2.3059842684391403,
  'labeledLatLngs': [{'label': 'display',
    'lat': 48.85943355462305,
    'lng': 2.3059842684391403}],
  'postalCode': '75015',
  'cc': 'FR',
  'city': 'Paris',
  'state': 'Île-de-France',
  'country': 'France',
  'formattedAddress': ['90 rue Saint-Dominique', '75015 Paris', 'France']},
 'canonicalUrl': 'https://foursquare.com/v/starbucks/4df1f8e13151247c51af4f21',
 'categories': [{'id': '4bf58dd8d48988d1e0931735',
   'name': 'Coffee Shop',
   'pluralName': 'Coffee Shops',
   'shortName': 'Coffee Shop',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/co

In [13]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

7.2


### 7.2 is a good rating. So, lets go there and have our coffee and some snacks.

## Explore a location

### Now after we had our coffee at starbucks, lets check out the popular spots around the cafe.

In [91]:
latitude = 48.859434
longitude = 2.305984

In [92]:
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=WVVDZRIEPUF43TCVXTB1T3JEHNC41AAS3YJAO4SRIQCE14U2&client_secret=CMB21XECJFAFCAZ5CDMP53AWTZBNUACTEBLGHSTBLPUQ2JFX&ll=48.859434,2.305984&v=20180604&radius=20000&limit=30'

In [93]:
import requests

In [102]:
results = requests.get(url).json()
'There are {} around starbucks.'.format(len(results['response']['groups'][0]['items']))

'There are 30 around starbucks.'

In [96]:
items = results['response']['groups'][0]['items']
items[0]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '4adcda13f964a520ad3621e3',
  'name': "L'Ami Jean",
  'location': {'address': '27 rue Malar',
   'lat': 48.860296563890564,
   'lng': 2.305888511863262,
   'labeledLatLngs': [{'label': 'display',
     'lat': 48.860296563890564,
     'lng': 2.305888511863262}],
   'distance': 96,
   'postalCode': '75007',
   'cc': 'FR',
   'city': 'Paris',
   'state': 'Île-de-France',
   'country': 'France',
   'formattedAddress': ['27 rue Malar', '75007 Paris', 'France']},
  'categories': [{'id': '4bf58dd8d48988d10c941735',
    'name': 'French Restaurant',
    'pluralName': 'French Restaurants',
    'shortName': 'French',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/french_',
     'suffix': '.png'},
    'primary': True}],
  'photos': {'count': 0, 'groups': []}},
 'referralId': 'e-0-4adcda13f964a520ad3621e3-0'}

In [97]:
dataframe = json_normalize(items) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

name           categories  \
0                                         L'Ami Jean    French Restaurant   
1                                Place des Invalides                Plaza   
2                                Hôtel Plaza Athénée                Hotel   
3                                        Tour Eiffel  Monument / Landmark   
4                                       Grand Palais        Historic Site   
5                              Jardin du Musée Rodin               Garden   
6  Hôtel National des Invalides – Musée de l'Armé...        Historic Site   
7                                     Christian Dior          Tailor Shop   
8                                    Palais de Tokyo           Art Museum   
9                         Théâtre des Champs-Élysées          Music Venue   

                          address  cc   city country  \
0                    27 rue Malar  FR  Paris  France   
1             Place des Invalides  FR  Paris  France   
2             25 avenue Montaigne  FR  Paris  France   
3         5 avenue Anatole France  FR  Paris  France   
4  3 avenue du Général Eisenhower  FR  Paris  France   
5               79 rue de Varenne  FR  Paris  France   
6         Esplanade des Invalides  FR  Paris  France   
7             30 avenue Montaigne  FR  Paris  France   
8   13 avenue du Président Wilson  FR  Paris  France   
9             15 avenue Montaigne  FR  Paris  France   

                crossStreet  distance  \
0                       NaN        96   
1                       NaN       517   
2         Rue Clément Marot       766   
3                       NaN       856   
4  Avenue Winston Churchill       888   
5               Musée Rodin       823   
6                       NaN       575   
7                       NaN       805   
8                       NaN       862   
9                       NaN       728   

                                    formattedAddress  \
0                [27 rue Malar, 75007 Paris, France]   
1         [Place des Invalides, 75007 Paris, France]   
2  [25 avenue Montaigne (Rue Clément Marot), 7500...   
3     [5 avenue Anatole France, 75007 Paris, France]   
4  [3 avenue du Général Eisenhower (Avenue Winsto...   
5  [79 rue de Varenne (Musée Rodin), 75014 Paris,...   
6     [Esplanade des Invalides, 75007 Paris, France]   
7         [30 avenue Montaigne, 75008 Paris, France]   
8  [13 avenue du Président Wilson, 75016 Paris, F...   
9         [15 avenue Montaigne, 75008 Paris, France]   

                                      labeledLatLngs        lat       lng  \
0  [{'label': 'display', 'lat': 48.86029656389056...  48.860297  2.305889   
1  [{'label': 'display', 'lat': 48.85863052623177...  48.858631  2.312944   
2  [{'label': 'display', 'lat': 48.86623539871467...  48.866235  2.304374   
3  [{'label': 'display', 'lat': 48.85824935460562...  48.858249  2.294426   
4  [{'label': 'display', 'lat': 48.86611219637595...  48.866112  2.312622   
5  [{'label': 'display', 'lat': 48.85583847334092...  48.855838  2.315810   
6  [{'label': 'display', 'lat': 48.85681640567485...  48.856816  2.312762   
7  [{'label': 'display', 'lat': 48.866666, 'lng':...  48.866666  2.306178   
8  [{'label': 'display', 'lat': 48.86420659410110...  48.864207  2.296700   
9  [{'label': 'display', 'lat': 48.8657, 'lng': 2...  48.865700  2.303133   

     neighborhood postalCode          state                        id  
0             NaN      75007  Île-de-France  4adcda13f964a520ad3621e3  
1             NaN      75007  Île-de-France  50c5c94690e732743b022d32  
2             NaN      75008  Île-de-France  4b817458f964a52024a730e3  
3             NaN      75007  Île-de-France  51a2445e5019c80b56934c75  
4             NaN      75008  Île-de-France  4c25f508a852c9281ab1e56c  
5             NaN      75014  Île-de-France  4dc790c01fc72e86376fc69c  
6             NaN      75007  Île-de-France  4adcda09f964a520fd3321e3  
7             NaN      75008  Île-de-France  4adcda1ff964a5205b3921e3  
8        Chaillot 

In [104]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Starbucks


# add Starbucks as a red circle mark
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='starbucks',
    fill=True,
    color='red',
    fill_color='red',
    fill_opacity=0.6
    ).add_to(venues_map)


# add popular spots to the map as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(venues_map)

# display map
venues_map

### Now we can select from a variety of options such as Jardin du Musée Rodin, or Théâtre des Champs-Élysées to spend some time. 

### Thank You!